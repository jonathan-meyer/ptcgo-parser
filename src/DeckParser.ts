import Deck from './Deck';
import { BASIC_ENERGY_IDS, SET_CODES } from './types';

const CARD_PATTERN = /^(?:\* )?(\d+) (.*) (\w+|[A-Z]{2}-[A-Z]{2})? (\d+|XY\d+|BW\d+)$/;

const SET_PATTERN = /(?:\* )?(\d+) (.*) ([A-Z]{2,3}|[A-Z]{2}-[A-Z]{2}|[A-Z0-9]{3})? (\d+|XY\d+|BW\d+)*/;
const BASIC_ENERGY_PATTERN = /(?:\* )?(\d+) (Darkness|Fairy|Fighting|Fire|Grass|Lightning|Metal|Psychic|Water) Energy.*/;

const parseRow = (row: string) => {
  const [, amount, name, set, code] = row.match(CARD_PATTERN) || [];
  return { amount, name, set, code };
};

class DeckParser {
  static parse(decklist: string): Deck {
    return {
      cards: decklist
        .split('\n')
        .map((row) => row.trim())
        .filter((row) => row.length > 0)
        .map((row) => {
          const { amount, name, set, code } = parseRow(row);

          const promoSet =
            set && set.startsWith('PR') ? set.split('-')[1] : null;

          return name
            ? {
                raw: row,
                amount: Number(amount),
                name,
                set,
                code: Number(code),
                ptcgoio: {
                  id: promoSet
                    ? `${SET_CODES[set]}-${promoSet}${code}`
                    : `${SET_CODES[set]}-${
                        Number(code) + (set === 'Energy' ? 163 : 0)
                      }`,
                },
              }
            : null;
        })
        .filter((card) => card),
    };
  }
}

export default DeckParser;
