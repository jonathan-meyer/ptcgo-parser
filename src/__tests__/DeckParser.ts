import { readFileSync } from 'fs';
import { resolve } from 'path';
import DeckParser from '../DeckParser';

describe('DeckParser', () => {
  it('Should Work with an empty string', () => {
    expect(DeckParser.parse('')).toEqual({ cards: [] });
  });

  it('Should Work #1', () => {
    expect(DeckParser.parse('* 1 Tapu Koko PR-SM 30')).toEqual({
      cards: [
        {
          amount: 1,
          code: 30,
          name: 'Tapu Koko',
          ptcgoio: {
            id: 'smp-SM30',
          },
          set: 'PR-SM',
        },
      ],
    });
  });

  it('Should Work #2', () => {
    expect(DeckParser.parse('* 1 Oranguru SUM 113')).toEqual({
      cards: [
        {
          amount: 1,
          code: 113,
          name: 'Oranguru',
          ptcgoio: {
            id: 'sm1-113',
          },
          set: 'SUM',
        },
      ],
    });
  });

  it('Should Work #3', () => {
    expect(DeckParser.parse('* 2 Mew FAC 29')).toEqual({
      cards: [
        {
          amount: 2,
          code: 29,
          name: 'Mew',
          ptcgoio: {
            id: 'xy10-29',
          },
          set: 'FAC',
        },
      ],
    });
  });

  it('Should Work #4', () => {
    expect(DeckParser.parse('* 11 Darkness Energy 7')).toEqual({
      cards: [
        {
          amount: 11,
          code: 7,
          name: 'Darkness',
          ptcgoio: {
            id: 'sm1-170',
          },
          set: 'Energy',
        },
      ],
    });
  });

  it('Should Work #5', () => {
    expect(DeckParser.parse('* 10 Lightning Energy 1')).toEqual({
      cards: [
        {
          amount: 10,
          code: 1,
          name: 'Lightning',
          ptcgoio: {
            id: 'sm1-167',
          },
          set: 'Energy',
        },
      ],
    });
  });

  it('Should Work #6', () => {
    expect(DeckParser.parse('* 4 Double Dragon Energy ROS 97')).toEqual({
      cards: [
        {
          amount: 4,
          code: 97,
          name: 'Double Dragon Energy',
          ptcgoio: {
            id: 'xy6-97',
          },
          set: 'ROS',
        },
      ],
    });
  });

  it('Should Work with an exported deck', () => {
    const deck = readFileSync(resolve(__dirname, 'test.deck')).toString();
    expect(DeckParser.parse(deck)).toMatchSnapshot();
  });
});

/*
const PTCGOParser = require('../lib/index.js')
const fs = require('fs')
const path = require('path')

const pick = (decklist, card_name) =>
  decklist.cards.filter(card => card.name === card_name)[0]

const exported_list = fs
  .readFileSync(path.resolve(__dirname, 'exported-1.deck'))
  .toString()

const promo_deck = fs
  .readFileSync(path.resolve(__dirname, 'promo.deck'))
  .toString()

const energies_deck = fs
  .readFileSync(path.resolve(__dirname, 'energies.deck'))
  .toString()

describe('ptcgo-parser', () => {
  it('should ignore headlines', () => {
    const decklist = PTCGOParser.parse(exported_list)
    const header_items = decklist.cards.filter(item =>
      item.name.startsWith('##')
    )

    expect(header_items.length).toBe(0)
  })

  it('should parse Oranguru correctly', () => {
    const decklist = PTCGOParser.parse(exported_list)
    const oranguru = pick(decklist, 'Oranguru')

    expect(oranguru.name).toBe('Oranguru')
    expect(oranguru.amount).toBe('1')
    expect(oranguru.set).toBe('SUM')
    expect(oranguru.code).toBe('113')
  })

  it('should work with cards without leading asterisk', () => {
    const decklist = PTCGOParser.parse(exported_list)
    const card = pick(decklist, 'Special Charge')

    expect(card.name).toBe('Special Charge')
    expect(card.amount).toBe('1')
    expect(card.set).toBe('STS')
    expect(card.code).toBe('105')
  })

  it('should parse correctly basic energy without set code', () => {
    const decklist = PTCGOParser.parse(exported_list)
    const card = pick(decklist, 'Darkness')

    expect(card.name).toBe('Darkness')
    expect(card.amount).toBe('11')
    expect(card.set).toBe(undefined)
    expect(card.code).toBe(undefined)
  })

  it('should parse correctly basic energy with a set code', () => {
    const decklist = PTCGOParser.parse(exported_list)
    const card = pick(decklist, 'Water')

    expect(card.name).toBe('Water')
    expect(card.amount).toBe('4')
    expect(card.set).toBe(undefined)
    expect(card.code).toBe(undefined)
  })

  it('should parse correctly special energy with a set code', () => {
    const decklist = PTCGOParser.parse(exported_list)

    const card = pick(decklist, 'Double Dragon Energy')

    expect(card.name).toBe('Double Dragon Energy')
    expect(card.amount).toBe('4')
    expect(card.set).toBe('ROS')
    expect(card.code).toBe('97')
  })

  it('should parse correctly SM promo', () => {
    const decklist = PTCGOParser.parse(promo_deck)
    const card = pick(decklist, 'Tapu Koko')

    expect(card.name).toBe('Tapu Koko')
    expect(card.amount).toBe('1')
    expect(card.set).toBe('PR-SM')
    expect(card.code).toBe('30')
  })

  it('should parse correctly XY promo', () => {
    const decklist = PTCGOParser.parse(promo_deck)
    const card = pick(decklist, 'Jirachi')

    expect(card.name).toBe('Jirachi')
    expect(card.amount).toBe('1')
    expect(card.set).toBe('PR-XY')
    expect(card.code).toBe('67')
  })

  it('should match PTCGO set code to ptcgo.io id correctly', () => {
    const decklist = PTCGOParser.parse(exported_list)
    const card = pick(decklist, 'Double Dragon Energy')

    expect(card.name).toBe('Double Dragon Energy')
    expect(card.set).toBe('ROS')
    expect(card.ptcgoio.id).toBe('xy6-97')
  })
})

it('should match promo PTCGO set code to ptcgo.io id correctly', () => {
  const decklist = PTCGOParser.parse(promo_deck)

  const card = pick(decklist, 'Jirachi')

  expect(card.name).toBe('Jirachi')
  expect(card.set).toBe('PR-XY')
  expect(card.ptcgoio.id).toBe('xyp-XY67')
})

it('should match promo PTCGO set code to ptcgo.io id correctly, part 2', () => {
  const decklist = PTCGOParser.parse(promo_deck)

  const card = pick(decklist, 'Tapu Koko')

  expect(card.name).toBe('Tapu Koko')
  expect(card.set).toBe('PR-SM')
  expect(card.ptcgoio.id).toBe('smp-SM30')
})

it('should use S&M energy ids for basic energies', () => {
  const decklist = PTCGOParser.parse(energies_deck)

  let card = pick(decklist, 'Darkness')
  expect(card.name).toBe('Darkness')
  expect(card.ptcgoio.id).toBe('sm1-170')

  card = pick(decklist, 'Fairy')
  expect(card.name).toBe('Fairy')
  expect(card.ptcgoio.id).toBe('sm1-172')

  card = pick(decklist, 'Fighting')
  expect(card.name).toBe('Fighting')
  expect(card.ptcgoio.id).toBe('sm1-169')

  card = pick(decklist, 'Fire')
  expect(card.name).toBe('Fire')
  expect(card.ptcgoio.id).toBe('sm1-165')

  card = pick(decklist, 'Grass')
  expect(card.name).toBe('Grass')
  expect(card.ptcgoio.id).toBe('sm1-164')

  card = pick(decklist, 'Lightning')
  expect(card.name).toBe('Lightning')
  expect(card.ptcgoio.id).toBe('sm1-167')

  card = pick(decklist, 'Metal')
  expect(card.name).toBe('Metal')
  expect(card.ptcgoio.id).toBe('sm1-171')

  card = pick(decklist, 'Psychic')
  expect(card.name).toBe('Psychic')
  expect(card.ptcgoio.id).toBe('sm1-168')

  card = pick(decklist, 'Water')
  expect(card.name).toBe('Water')
  expect(card.ptcgoio.id).toBe('sm1-166')
})
*/
