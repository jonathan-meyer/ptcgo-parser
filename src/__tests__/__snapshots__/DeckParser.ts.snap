// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeckParser Should Work with an exported deck 1`] = `
Object {
  "cards": Array [
    Object {
      "amount": 1,
      "code": 113,
      "name": "Oranguru",
      "ptcgoio": Object {
        "id": "sm1-113",
      },
      "raw": "* 1 Oranguru SUM 113",
      "set": "SUM",
    },
    Object {
      "amount": 3,
      "code": 74,
      "name": "Darkrai-EX",
      "ptcgoio": Object {
        "id": "xy9-74",
      },
      "raw": "* 3 Darkrai-EX BKP 74",
      "set": "BKP",
    },
    Object {
      "amount": 4,
      "code": 94,
      "name": "Dratini",
      "ptcgoio": Object {
        "id": "sm1-94",
      },
      "raw": "* 4 Dratini SUM 94",
      "set": "SUM",
    },
    Object {
      "amount": 3,
      "code": 95,
      "name": "Dragonair",
      "ptcgoio": Object {
        "id": "sm1-95",
      },
      "raw": "* 3 Dragonair SUM 95",
      "set": "SUM",
    },
    Object {
      "amount": 1,
      "code": 51,
      "name": "Dragonite",
      "ptcgoio": Object {
        "id": "xy6-51",
      },
      "raw": "* 1 Dragonite ROS 51",
      "set": "ROS",
    },
    Object {
      "amount": 1,
      "code": 29,
      "name": "Mew",
      "ptcgoio": Object {
        "id": "xy10-29",
      },
      "raw": "* 1 Mew FCO 29",
      "set": "FCO",
    },
    Object {
      "amount": 1,
      "code": 99,
      "name": "Fighting Fury Belt",
      "ptcgoio": Object {
        "id": "xy9-99",
      },
      "raw": "* 1 Fighting Fury Belt BKP 99",
      "set": "BKP",
    },
    Object {
      "amount": 2,
      "code": 69,
      "name": "Ace Trainer",
      "ptcgoio": Object {
        "id": "xy7-69",
      },
      "raw": "* 2 Ace Trainer AOR 69",
      "set": "AOR",
    },
    Object {
      "amount": 4,
      "code": 92,
      "name": "Trainers' Mail",
      "ptcgoio": Object {
        "id": "xy6-92",
      },
      "raw": "* 4 Trainers' Mail ROS 92",
      "set": "ROS",
    },
    Object {
      "amount": 2,
      "code": 122,
      "name": "Lillie",
      "ptcgoio": Object {
        "id": "sm1-122",
      },
      "raw": "* 2 Lillie SUM 122",
      "set": "SUM",
    },
    Object {
      "amount": 3,
      "code": 146,
      "name": "Professor's Letter",
      "ptcgoio": Object {
        "id": "xy8-146",
      },
      "raw": "* 3 Professor's Letter BKT 146",
      "set": "BKT",
    },
    Object {
      "amount": 1,
      "code": 149,
      "name": "Super Rod",
      "ptcgoio": Object {
        "id": "xy8-149",
      },
      "raw": "* 1 Super Rod BKT 149",
      "set": "BKT",
    },
    Object {
      "amount": 2,
      "code": 78,
      "name": "Lysandre",
      "ptcgoio": Object {
        "id": "xy7-78",
      },
      "raw": "* 2 Lysandre AOR 78",
      "set": "AOR",
    },
    Object {
      "amount": 3,
      "code": 107,
      "name": "Professor Sycamore",
      "ptcgoio": Object {
        "id": "xy9-107",
      },
      "raw": "* 3 Professor Sycamore BKP 107",
      "set": "BKP",
    },
    Object {
      "amount": 1,
      "code": 77,
      "name": "Lucky Helmet",
      "ptcgoio": Object {
        "id": "xy7-77",
      },
      "raw": "* 1 Lucky Helmet AOR 77",
      "set": "AOR",
    },
    Object {
      "amount": 1,
      "code": 38,
      "name": "Switch",
      "ptcgoio": Object {
        "id": "xy0-38",
      },
      "raw": "* 1 Switch KSS 38",
      "set": "KSS",
    },
    Object {
      "amount": 1,
      "code": 136,
      "name": "Fisherman",
      "ptcgoio": Object {
        "id": "xy8-136",
      },
      "raw": "* 1 Fisherman BKT 136",
      "set": "BKT",
    },
    Object {
      "amount": 1,
      "code": 127,
      "name": "Escape Rope",
      "ptcgoio": Object {
        "id": "xy5-127",
      },
      "raw": "* 1 Escape Rope PRC 127",
      "set": "PRC",
    },
    Object {
      "amount": 1,
      "code": 105,
      "name": "Special Charge",
      "ptcgoio": Object {
        "id": "xy11-105",
      },
      "raw": "* 1 Special Charge STS 105",
      "set": "STS",
    },
    Object {
      "amount": 1,
      "code": 109,
      "name": "VS Seeker",
      "ptcgoio": Object {
        "id": "xy4-109",
      },
      "raw": "* 1 VS Seeker PHF 109",
      "set": "PHF",
    },
    Object {
      "amount": 3,
      "code": 113,
      "name": "Ultra Ball",
      "ptcgoio": Object {
        "id": "xy10-113",
      },
      "raw": "* 3 Ultra Ball FCO 113",
      "set": "FCO",
    },
    Object {
      "amount": 2,
      "code": 137,
      "name": "Float Stone",
      "ptcgoio": Object {
        "id": "xy8-137",
      },
      "raw": "* 2 Float Stone BKT 137",
      "set": "BKT",
    },
    Object {
      "amount": 2,
      "code": 76,
      "name": "Level Ball",
      "ptcgoio": Object {
        "id": "xy7-76",
      },
      "raw": "* 2 Level Ball AOR 76",
      "set": "AOR",
    },
    Object {
      "amount": 1,
      "code": 183,
      "name": "Rainbow Energy",
      "ptcgoio": Object {
        "id": "sm7-183",
      },
      "raw": "* 1 Rainbow Energy CES 183",
      "set": "CES",
    },
    Object {
      "amount": 4,
      "code": 97,
      "name": "Double Dragon Energy",
      "ptcgoio": Object {
        "id": "xy6-97",
      },
      "raw": "* 4 Double Dragon Energy ROS 97",
      "set": "ROS",
    },
    Object {
      "amount": 10,
      "code": 7,
      "name": "Darkness Energy",
      "ptcgoio": Object {
        "id": "sm1-170",
      },
      "raw": "* 10 Darkness Energy Energy 7",
      "set": "Energy",
    },
    Object {
      "amount": 1,
      "code": 3,
      "name": "Water Energy",
      "ptcgoio": Object {
        "id": "sm1-166",
      },
      "raw": "* 1 Water Energy Energy 3",
      "set": "Energy",
    },
    Object {
      "amount": 18,
      "code": 8,
      "name": "Psychic Energy",
      "ptcgoio": Object {
        "id": "undefined-8",
      },
      "raw": "* 18 Psychic Energy XYEnergy 8",
      "set": "XYEnergy",
    },
  ],
}
`;
